Hlib bondarev/store draft in cache

HlibBondarev opened the pull request:
**Roles:** provider, ministryadmin, areaadmin, regionadmin, techadmin.

**Story:** As a User with the above roles, I want to be able to save the field values before saving the Workshop (Provider, etc.) in the DB. So that if the  Workshop (Provider, etc.) creation is terminated, I can restore the values of previously filled fields.

**Solution**: Created DraftStorageService.  It captures and exports the state of an object (entity) outside of it so that the state of the original object can be restored later. Redis was used to store intermediate values of entity fields.

1) Created IDraftStorageService<T> interface and implemented it in StorageService<T> class .
2) Created IReadWriteCacheService interface with methods - ReadAsync(), WriteAsync() and RemoveAsync().
3) Implemented IReadWriteCacheService interface in CacheService class.
4) Refactored  CacheService class - used ReadAsync() and WriteAsync()  methods in GetOrAddAsync<T>() method.
5) Fixed RemoveAsync() method - instead of cache.RemoveAsync(key) used cache.Remove(key).
6) Created abstract generic controller DraftStorageController<T> with actions StoreDraft, RestoreDraft and RemoveDraft.
7) Created WorkshopDraftStorageController inherited from DraftStorageController<T> with T = WorkshopBaseDto.
8) Added tests to CacheServiceTests class for methods - ReadAsync(), WriteAsync() and RemoveAsync().
9) Created DraftStorageServiceTests class with NUnit tests for DraftStorageService class methods - RestoreAsync(), CreateAsync() and RemoveAsync().

HlibBondarev committed:
1) Created IRedisCacheService interface with methods GetValueFromRedisCacheAsync(), SetValueToRedisCacheAsync() and RemoveValueFromRedisCacheAsync();
2) Implemented IRedisCacheService in CacheService();
3) Created IMemento, IMementoService<T> and IStorage interfaces and imolemented them in classes - Memento, MementoService<T> and Storage.
4) Created abstract controller BaseMementoController<T> with actions StoreMemento, RestoreMemento and RemoveMemento.
5) Created WorkshopMementoController inherited from BaseMementoController<T> whit T = RequiredWorkshopMemento.

HlibBondarev committed:
Refactored CacheService class:
1) Renamed IRedisCacheService to ICrudCacheService;
2) Renamed methods in ICrudCacheService (new names - GetValueFromCacheAsync(), SetValueToCacheAsync() and RemoveFromCacheAsync());
3) Changed methods GetOrAddAsync<T>() (used GetValueFromCacheAsync(), SetValueToCacheAsync() in it) and RemoveAsync() (used method cache.Remove(key) instead of cache.RemoveAsync(key)).

HlibBondarev committed:
Removed the redundant code line in StatisticReportController class.

HlibBondarev committed:
Removed redundant Using directive in StatisticReportController class.

HlibBondarev committed:
Added Rolles to attribute [Authorize(Roles = 'provider, ministryadmin, areaadmin, regionadmin, techadmin')] for methods StoreMemento(), RestoreMemento() and RemoveMemento() in BaseMementoController<T>.

HlibBondarev committed:
1) Removed logging from MementoService class.
2) Changed logging in Storage class - added throwing InvalidOperationException(errMessage).
3) Fixed the logic of GetValueFromCacheAsync method in CacheService class.

HlibBondarev committed:
Changed the return string in actions - StoreMemento and RemoveMemento of BaseMementoController<T>.

HlibBondarev committed:
Added DataAnnotation attributes to RequiredWorkshopMemento entity for validation its properties.

HlibBondarev committed:
Documented code for classes - Memento, MementoService, Storage, BaseMementoController and WorkshopMementoController.

HlibBondarev committed:
Fixed test - RemoveAsync_ShouldCallCacheRemove() in CacheServiceTests class.

HlibBondarev committed:
Fixed test - RemoveAsync_ShouldCallCacheRemove() in CacheServiceTests class.

HlibBondarev committed:
Fixed GetValueFromCacheAsync() method in in CacheService class.

HlibBondarev committed:
Added tests to CacheServiceTests class for methods - GetValueFromCacheAsync(), SetValueToCacheAsync() and RemoveFromCacheAsync().

HlibBondarev committed:
Created StorageTests class with NUnit tests for Statorage class methods - SetMementoValueAsync(), GetMementoValueAsync() and RemoveMementoAsync().

HlibBondarev committed:
Created MementoServiceTests class with NUnit tests for MementoService class methods - RestoreMemento(), CreateMemento() and GetMementoKey().

HlibBondarev committed:
Fixed Sonar Issues.

HlibBondarev committed:
Deleted redundant DataAnnotation attributes in RequiredWorkshopMemento class.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/Storage.cs:
```diff
@@ -0,0 +1,81 @@
+    public async Task<KeyValuePair<string, string?>> GetMementoValueAsync(string key)
+    {
+        string logMsg = $"Getting memento with key = {key} from cache has started.";
+        logger.LogDebug(logMsg);
```
(Logging too verbose due to wrong level) Not sure if this rather technical info should pe part of the "Debug" logs. It looks more like Information/Trace level

HlibBondarev replied:
IStorage and Storage has been removed from code.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/Storage.cs:
```diff
@@ -0,0 +1,81 @@
+        if (value == null)
+        {
+            logMsg = $"Memento with key = {key} not found in cache.";
+            logger.LogWarning(logMsg);
```
(Log level) Is empty cache an issue or a normal case? I believe it's latter. Then why show warning? It's Debug/Info/Trace level, not higher

HlibBondarev replied:
IStorage and Storage has been removed from code.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Redis/CacheService.cs:
```diff
@@ -8,26 +8,25 @@
-
     private readonly object lockObject = new object();
-    
+    bool isExists = false;
```
(State in srvices) Please avoid usage of any kind of state in services, as they should support shared access. 
The issue is next: serivce state (`isExists`) is shared for all service methods but is related to specific method input (`string key`). Thus I can in parallel call the `GetOrAddAsync()` twice or together with `GetValueFromCacheAsync()` (both are working with `isExists`), passing there different keys. What `key` this `isExists` will be related as a result, I dunno

HlibBondarev replied:
isExists  field is removed from CacheService class.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Redis/CacheService.cs:
```diff
@@ -46,73 +45,92 @@ public async Task<T> GetOrAddAsync<T>(
+        }
+
+        return returnValue;
+    }
```
Here you've broken previously existing logic of not adding same data twice.
Imagine `GetOrAddAsync` is called at same time for for same `key`, that is absent in cache.
You'll get `SetValueToCacheAsync` called twice in result. 
It's 
- either ok, then you don't need `cacheLock.EnterReadLock()` in `GetValueFromCacheAsync()`
- or wrong, then your code should be rewritten

HlibBondarev replied:
I returned the logic that was before the changes in the code.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Redis/CacheService.cs:
```diff
@@ -46,73 +45,92 @@ public async Task<T> GetOrAddAsync<T>(
         });
+    }
+
+    public Task RemoveFromCacheAsync(string key) => RemoveAsync(key);
```
Please clarify what is the purpose of this method.
If it is just to provide other name for same logic as `RemoveAsync` do, than it's bit excessive, as for **Cache**Service.**Remove**Async(key) it's obvious to remove from cache, and you don't need specify it as **Cache**Service.Remove**FromCache**Async(key)

HlibBondarev replied:
RemoveFromCacheAsync() is removed from CacheService class.
The RemoveAsync() method has been added to ICrudCacheService so that it can be used in MementoService where the ICrudCacheService interface is used.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Redis/ICrudCacheService.cs:
```diff
@@ -0,0 +1,17 @@
+
+public interface ICrudCacheService
+{
+    Task<string?> GetValueFromCacheAsync(string key);
```
( Naming again) ICrud**Cache**Service already has `Cache` in name, don't need to repeat it in all methods, like GetValue**FromCache**Async(), SetValue**ToCache**Async(), Remove**FromCache**Async(). 
It's basic property of encapsulation - to specify your main object of interest on class/interface name level and work with it on methods level.

HlibBondarev replied:
Renamed methods in ICrudCacheService - removed the word 'Cache' from method names.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopMementoControllerTests.cs:
```diff
@@ -0,0 +1,204 @@
+    //    // Arrange
+    //    var data = new SearchResult<DirectionDto>()
+    //    {
+    //        Entities = new List<DirectionDto>() { new DirectionDto() },
```
(Commented code) Please don't commit commented code. Usually it's useless later (due to subsequent code/requirement changes), thus could be removed in initial commit

HlibBondarev replied:
This code is removed

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/Interfaces/IStorage.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace OutOfSchool.BusinessLogic.Services.Memento.Interfaces;
+
+public interface IStorage
```
(Naming) Interface pretend to be a **Storage**. All methods are working with **Memento**. Looks bit irrelevant.
My recommendation - remove `Memento` part from all methods' names.

VadymLevkovskyi replied:
Also, see other comment in `BaseMementoController`. Looks like this `IStorage` could be joined with memento service.

HlibBondarev replied:
IStorage and Storage has been removed from code.
Also IMemento and Memento has been removed from code.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/Interfaces/IMementoService.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    T? State { get; set; }
+
+    IMemento CreateMemento(string key, T value);
```
(Naming) Too many **Memento**s. Suggestion is to not duplicate information that is already set on class/interface name level (that it works with memento) in names of methods

HlibBondarev replied:
Renamed methods in MementoService - removed the word 'memento' from method names.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,61 @@
+
+    /// <summary>Gets or sets the State of the memento.</summary>
+    /// <value>The state of the memento.</value>
+    public T? State { get; set; }
```
(Shared state in services) Issue similar to `isExists` mentioned in other comments. You could change it by parallel calling `RestoreMemento(memento)` with different input and get unpredictable result in `State`.

Suggestion - remove `State` propertty at all. Instead return the `State` data from `RestoreMemento()`.

HlibBondarev replied:
State property is removed from MementoService class.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,61 @@
+
+    /// <summary>Restores the memento.</summary>
+    /// <param name="memento">The memento.</param>
+    public void RestoreMemento(KeyValuePair<string, string?> memento)
```
(Probably wrong type) `KeyValuePair<string, string?> memento` - this inut type is very similar to `IMemento`. What is the reason to not use it?

HlibBondarev replied:
IMemento and Memento has been removed from code.
Method signature changed to Task<T> RestoreAsync(string key).

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <returns>
+    /// Memento key of string type.
+    /// </returns>
+    public string GetMementoKey(string key)
```
(Class design) I don't see reason to have this GetMementoKey() as a public method. In each call site you call it an then some other method form the same `MementoService`. Why don't make `GetMementoKey()` private one and call it inside `MementoService` public methods (like you do in `CreateMemento()`)?

HlibBondarev replied:
It's done. 
GetMementoKey() method has been made private.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/Storage.cs:
```diff
@@ -0,0 +1,81 @@
+/// <summary>
+/// Implements the IStorage interface to get, set, and remove memento in the cache.
+/// </summary>
+public class Storage : IStorage
```
(Occam's razor) This whole `Storage` implementation looks like logging wrapper over `ICrudCacheService`. Why don't do all this things inside `ICrudCacheService` and don't introduce another entity with limited logic?

HlibBondarev replied:
It's done. 
Storage has been removed from code.
All CRUD logic of cache is relocated to ICrudCacheService.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/BaseMementoController.cs:
```diff
@@ -0,0 +1,75 @@
+        IMementoService<T> mementoService,
+        IStorage storage)
+    {
+        this.mementoService = mementoService ?? throw new ArgumentNullException(nameof(mementoService));
+        this.storage = storage ?? throw new ArgumentNullException(nameof(storage));
```
(Logic) Controllers should usually avoid containing logic. In this controller you have 2 services and introduce logic of calling both of them in specific order. As for me it looks like they are highly coupled, that shouldn't be (remember, _high cohesion, low coupling_). 
Probably something is wrong on design level and these 2 services should be the one (or at least hidden by the facade of one additional service)?

HlibBondarev replied:
It's done. 
Storage has been removed from code.
All CRUD logic of cache is relocated to ICrudCacheService.

VadymLevkovskyi reviewed (changes requested):
Naming, logic, design - please see other comments. All tests weren't reviewed, as they probably will be rewritten.

HlibBondarev committed:
Removed service state (is Exists) from CacheService class.

HlibBondarev committed:
Changed CacheService methods

HlibBondarev committed:
Changed methods in CacheService and CacheServiceTests

HlibBondarev committed:
{

HlibBondarev committed:
1) Renamed methods in MementoService - removed the word 'memento' from method names.
2) Added validation (ModelState.IsValid) in StoreMemento action of BaseMementoController<T>.
3) Added NUnit tests to WorkshopMementoControllerTests class for method - StoreMemento().

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,81 @@
+    /// <returns>Representation of an asynchronous operation - removing memento from the cache.</returns>
+    public async Task RemoveAsync(string key)
+    {
+        string logMessage;
```
(Variable declaration placement) There are 2 suggestions I have about variable declarations:
1. Put them as close to usage as possible (not at the method beginning, at least)
2. Declare variable with immediate initialization - thus you could use 'var' instead of explicit type.

HlibBondarev replied:
It's done.
Declared a variable with immediate initialization in the RemoveAsync() method of the MementoService<T> class.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,81 @@
+
+        if (valueToRemove == null)
+        {
+            logMessage = $"Memento with key = {mementoKey} was not found in the cache.";
```
What is the reason to not pass this string into `LogInformation()` directly without variable?

HlibBondarev replied:
I passed the interpolated string to LogInformation() directly, without a variable. But Sonar wrote this warning - Don't use string interpolation in logging message templates.

https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&sinceLeakPeriod=true&pullRequest=1551&id=ita-social-projects-oos-backend

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,81 @@
+
+    private static string GetMementoKey(string key)
+    {
+        return string.Format("{0}_{1}", key, typeof(T).Name);
```
(Code style) Why not use string interpolation here? (That $"" syntax)

HlibBondarev replied:
It's done.
Replaced String.Format with interpolation in the GetMementoKey() method of the MementoService<T> class.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Redis/CacheService.cs:
```diff
@@ -95,24 +93,71 @@ await ExecuteRedisMethod(() =>
+        TimeSpan? absoluteExpirationRelativeToNowInterval = null,
+        TimeSpan? slidingExpirationInterval = null)
+    {
+        await ExecuteRedisMethod(() =>
+        {
```
(Code duplication) This method looks like a copy of part of the `GetOrAdd()` implementation. 
If it is so and to avoid duplication could it be called in `GetOrAdd()` instead of existing code?
If it is not a duplication - then why? For me it seems to do same thing

HlibBondarev replied:
Today, I've had a discussion about MementoService and CacheService with @h4wk13.
He suggested not to create a new entity (RequiredWorkshopMemento) and then serialize it and pass it to Cache. Instead, work directly with the Json-string and store it into the cache. Then all work will be done in CacheService class.
I asked this question to @DmyMi and now I'm wating the answer.

**UPD**
I had a talk with  @DmyMi on Friday.
He explained how the code should work, and why you need to pass a DTO entity to the Controller Action, not just a string. I want to discuss this with @h4wk13 on Monday

HlibBondarev replied:
> (Code duplication) This method looks like a copy of part of the `GetOrAdd()` implementation. If it is so and to avoid duplication could it be called in `GetOrAdd()` instead of existing code? If it is not a duplication - then why? For me it seems to do same thing

I agree with you that the code in GetOrAddAsync<T>() method is similar to the code in new methods.
In my opinion, this happened because this method should not be related to the ICacheService interface, since it has a query to the database. 
I think that a new interface should be created (I don’t know what to call it yet) in which the GetOrAddAsync<T>() method will be included. And all other methods (RemoveAsync(), GetValueAsync() and UpsertValueAsync<T>()) have to be in ICasheService interface. Then ICrudCacheService does not make sense and should be removed.

VadymLevkovskyi replied:
Sorry, where we have query to db in this code?

HlibBondarev replied:
parameter - Func<Task<T>> newValueFactory,
if (!isExists)
{
    returnValue = await newValueFactory();
...

VadymLevkovskyi replied:
I get it finally. It's not a "query to db" its just a "value factory". And I see the difference now.
But I still don't see why it's impossible to reduce code duplication. You can use `Upsert` inside `GetOrAdd`.

HlibBondarev replied:
Ok, I used ReadAsync() and WriteAsync inside GetOrAdd.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopMementoController.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    /// <summary>Initializes a new instance of the <see cref="WorkshopMementoController" /> class.</summary>
+    /// <param name="mementoService">The memento service.</param>
+    public WorkshopMementoController(
```
(Code style) Primary constructor could be used here

HlibBondarev replied:
It's done.
Used the primary constructor in the WorkshopMementoController.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,81 @@
+    public async Task RemoveAsync(string key)
+    {
+        string logMessage;
+        string mementoKey = GetMementoKey(key);
+        var valueToRemove = await crudCacheService.GetValueAsync(mementoKey);
```
(Code style) Please be consistent about `var` usage. Either don't use it at all or use it everywhere. Here in 2 lines of code variables are defined in 2 different ways

HlibBondarev replied:
It's done.
Replaced the variable type 'string' with the keyword 'var'.

VadymLevkovskyi reviewed (commented):
Code style comments

HlibBondarev committed:
1) Declared a variable with immediate initialization in the RemoveAsync() method of the MementoService<T> class.
2) Passed a string with interpolation to LogInformation() directly without a variable.
3) Replaced String.Format with interpolation in the GetMementoKey() method of the MementoService<T> class.
4)
5) Used the primary constructor in the WorkshopMementoController.
6) Replaced the variable type 'string' with the keyword 'var'.
7) Added validation of parameters passed to  RestoreAsync(),  CreateAsync() and RemoveAsync() methods of MementoService class.

HlibBondarev committed:
Fixed message for ArgumentNullException in RestoreAsync(), CreateAsync() and RemoveAsync() methods of MementoService class.

HlibBondarev committed:
Fixed message for ArgumentNullException in RestoreAsync(), CreateAsync() and RemoveAsync() methods of MementoService class.

HlibBondarev committed:
Added '.ConfigureAwait(false)' to async methods in classes - MementoService and BaseMementoController<T>. If continueOnCapturedContext=false, code not to attempt to marshal the continuation back to the original context captured.

HlibBondarev committed:
1) GetValueAsync<T>() method was renamed to UpsertValueAsync<T>() in the CacheService class.
2) RequiredWorkshopMemento class was renamed to IncomplitedWorkshopDto.

HlibBondarev committed:
1) Changed signature of UpsertValueAsync() method in CrudCacheService interface and CacheService class.
2) Changed CreateAsync() method in  MementoService<T> class.
3) Changed appropriated tests in CacheServiceTests and MementoServiceTests classes.

HlibBondarev committed:
Corrected name to IncompletedWorkshopDto

HlibBondarev committed:
Added DTO models - WorkshopWithRequiredPropertiesDto, WorkshopWithDescriptionDto, WorkshopWithContactsDto and WorkshopWithTeachersDto.

HlibBondarev committed:
Removed IncompletedWorkshopDto class from project.

HlibBondarev committed:
Changed WorkshopBaseDto class.

HlibBondarev committed:
Implemented Polymorphic JSON serialisation in Web API.
1) Created JsonCreationConverter<T> and WorkshopConverter for Polymorphic JSON serialisation.
2) Added [JsonDerivedType] Attribute for all derived classes to base class - WorkshopWithRequiredPropertiesDto.
3) Used JsonSerializer from System.Text.Json namespace in MementoService<T> class.
4) Added support of WorkshopConverter to Controllers in Startup class.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/Interfaces/IMementoService.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace OutOfSchool.BusinessLogic.Services.Memento.Interfaces;
+public interface IMementoService<T>
```
Why do we want to state that this is a Memento ?
Is that business logic related ?

HlibBondarev replied:
Yes, I think that we store in Redis not data but some notes (memento)

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,82 @@
+    /// <returns> Representing the asynchronous operation with result of T type.</returns>
+    public async Task<T> RestoreAsync([NotNull] string key)
+    {
+        var mementoKey = key is not null ? GetMementoKey(key) : throw new ArgumentNullException(nameof(key));
```
1. please move arguments validation logic out from the business logic.
2. Please use `ArgumentException.ThrowIfNullOrEmpty(nameof(key))` instead.

It will bring more readability to the code.
Update all other places.

HlibBondarev replied:
> 1. please move arguments validation logic out from the business logic.
> 2. Please use `ArgumentException.ThrowIfNullOrEmpty(nameof(key))` instead.
> 
> It will bring more readability to the code. Update all other places.

1) When we discussed last time this code I understood that I had to check all arguments that input to service methods.

2) Ok, I changed code

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,82 @@
+    {
+        var mementoKey = key is not null ? GetMementoKey(key) : throw new ArgumentNullException(nameof(key));
+
+        var memento = await crudCacheService.GetValueAsync(mementoKey).ConfigureAwait(false);
```
What exactly `memento` means here ?
We're storing instance of the memento pattern in the storage ? Why ?

HlibBondarev replied:
I think that we store in Redis not data but some notes (memento)

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,82 @@
+/// <typeparam name="T">T is the entity type that should be stored in the cache.</typeparam>
+public class MementoService<T> : IMementoService<T>
+{
+    private readonly ICrudCacheService crudCacheService;
```
`Crud` ? in name ? why ?

HlibBondarev replied:
There are all CRUD operations in ICrudCacheService - Create and Update (UpsertValueAsync), Read (GetValueAsync) and Delete (RemoveAsync).


Add my discussion with [VadymLevkovskyi](https://github.com/VadymLevkovskyi) about Code duplication in CasheService class:

> > (Code duplication) This method looks like a copy of part of the `GetOrAdd()` implementation. If it is so and to avoid duplication could it be called in `GetOrAdd()` instead of existing code? If it is not a duplication - then why? For me it seems to do same thing
> 
> I agree with you that the code in GetOrAddAsync() method is similar to the code in new methods. In my opinion, this happened because this method should not be related to the ICacheService interface, since it has a query to the database. I think that a new interface should be created (I don’t know what to call it yet) in which the GetOrAddAsync() method will be included. And all other methods (RemoveAsync(), GetValueAsync() and UpsertValueAsync()) have to be in ICasheService interface. Then ICrudCacheService does not make sense and should be removed.

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        logger.LogInformation($"Removing memento with key = {mementoKey} from cache has started.");
+        await crudCacheService.RemoveAsync(mementoKey).ConfigureAwait(false);
```
is it possible that value will be removed from the cache between that `if and this line ?

HlibBondarev replied:
Thanks, I agree.
I deleted code line - 
logger.LogInformation($"Memento with key = {mementoKey} has been removed from the cache.");

h4wk13 commented on OutOfSchool/OutOfSchool.Redis/CacheService.cs:
```diff
@@ -95,24 +95,71 @@ await ExecuteRedisMethod(() =>
 
     public Task RemoveAsync(string key)
-        => ExecuteRedisMethod(async () => {
+        => this.ExecuteRedisMethod(() =>
```
Could you please look inside this method and check if we can stuck in the internal `while` for a wile ?
Also please, invert the `if` inside it,

HlibBondarev replied:
I can't find _if_ in this method

h4wk13 commented on OutOfSchool/OutOfSchool.Redis/CacheService.cs:
```diff
@@ -95,24 +95,71 @@ await ExecuteRedisMethod(() =>
+    {
+        string returnValue = string.Empty;
+
+        await ExecuteRedisMethod(() =>
```
I see that Enter/Exit... in many places.
Is that possible to create 2 methods to wrap that logic ?
One for read and one for write locks ?

vpustovoitov replied:
Is locking necessary within this service? Since `CacheService` is either a **Scoped** or **Transient** service, concurrency shouldn’t be an issue.

In my opinion, for these requirements, we could create a service that isn’t coupled with the database, acting as a temporary data storage solution.

Alternatively, we could create a **Facade** for interacting with **Redis** and a **Decorator** for caching database responses. This way, we could use the **Facade** service to store temporary data.

UPD: Oh, as I see, `CacheService` is injected as a **Singleton** service (but why ?).

VadymLevkovskyi replied:
Ok, we have `DraftStorageService` and `IDistributedCache` and `CacheService`. "Not that we needed all that for the trip" ([Fear and Loathing in Las Vegas](https://www.goodreads.com/quotes/59592-we-had-two-bags-of-grass-seventy-five-pellets-of-mescaline)) but obviously we could add another Facade

HlibBondarev replied:
> `IDistributedCache`

To implement the DraftStorageService<T> logic, I only need two methods - ReadAsync() and WriteAsync(). I could create this class as a facade over IDistributedCache. But in this case, in order for correctly reading and writing to the cache, I will need to implement the locking logic, which is already implemented in the methods of the CacheService class.
Therefore, I added the IReadWriteCacheService interface and implemented its methods in CacheService.
DraftStorageService<T> class is designed  as a facade over CacheService. This class works with entities of type T, unlike CacheService, which saves and restores information in string format.

h4wk13 commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/BaseMementoController.cs:
```diff
@@ -0,0 +1,61 @@
+
+/// <summary>Base Controller with operations for storing data in cache.</summary>
+/// <typeparam name="T">T is the entity type that should be stored in the cache.</typeparam>
+public abstract class BaseMementoController<T> : ControllerBase
```
Please rename it!

HlibBondarev replied:
Is the name MementoController<T> appropriated?

h4wk13 commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopMementoController.cs:
```diff
@@ -0,0 +1,14 @@
+[ApiController]
+[ApiVersion("1.0")]
+[Route("api/v{version:apiVersion}/[controller]/[action]")]
+public class WorkshopMementoController(IMementoService<WorkshopWithRequiredPropertiesDto> mementoService)
```
Are you sure it should contain a `memento` as a part of the name ?

HlibBondarev replied:
I think controller name should contain **memento** in it, because in API may contain another controller with **memento**, for example - ProviderMementoController

HlibBondarev committed:
Used 'ArgumentException.ThrowIfNullOrEmpty(nameof(key))' instead of 'throw new ArgumentNullException(nameof(key))'

HlibBondarev committed:
Used 'ArgumentException.ThrowIfNullOrEmpty(nameof(key))' instead of 'throw new ArgumentNullException(nameof(key))'

HlibBondarev committed:
Deleted code line with LogInformation about removing memento from the cache in RemoveAsync() method of MementoService<T> class.

HlibBondarev committed:
Renamed BaseMementoController<T> to MementoController<T>.

HlibBondarev committed:
Renamed ICrudCacheService to IReadWriteCacheService and renamed appropriated methods.

HlibBondarev committed:
Used 'ArgumentException.ThrowIfNullOrEmpty(arg, nameof(arg))' instead of 'throw new ArgumentNullException(nameof(arg))' in MementoService<T>, JsonCreationConverter<T> and WorkshopConverter.

HlibBondarev committed:
Removed the optional argument from the ArgumentException.ThrowIfNullOrEmpty method call.

HlibBondarev committed:
Fixed tests in MementoServiceTests and WorkshopMementoControllerTests

HlibBondarev committed:
Added tests to the MementoServiceTests class to test WorkshopWithDescriptionDto, WorkshopWithContactsDto and WorkshopWithTeachersDto.

HlibBondarev committed:
Fixed RemoveAsync() method in CacheService class - removed 'async' from delegate.

HlibBondarev committed:
Refactored CacheService class - used ReadAsync() and WriteAsync() methods in GetOrAddAsync<T>() method.

HlibBondarev committed:
Changed GetOrAddAsync<T>() method in CacheService class.

HlibBondarev committed:
Fixed attribute for RemoveMemento() action in MementoController<T> - [HttpDelete] instead of [HttpGet].

olsavchenk commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/JSONConverter/JSONCreationConverter.cs:
```diff
@@ -0,0 +1,44 @@
+
+public abstract class JsonCreationConverter<T> : JsonConverter
+{
+    public override bool CanWrite
```
(Code style) I think it will be more readable if you make it
`public override bool CanWrite { get => false; }`

HlibBondarev replied:
O.k. I changed code - used an exspression body for this method and CanConvert().

olsavchenk commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/JSONConverter/JSONCreationConverter.cs:
```diff
@@ -0,0 +1,44 @@
+        return target;
+    }
+
+    public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
```
(Code style) Isn't it better to just not add this method if you don't implement it?

HlibBondarev replied:
WriteJson() is an abstract method of JsonConverter class, so I have to implemented it.

olsavchenk replied:
Yes, you must implement abstract methods when inheriting abstract classes, but JsonCreationConverter is also abstract, so there is no need to override it (especially with a simple throwing error). JsonCreationConverter simply extends the JsonConverter's abstaction(if I may put it that way)

HlibBondarev replied:
If I don't implement WriteJson() method in JSONCreationConverter class, then I have to implement it in WorkshopConverter class, which is not very good in my opinion. The only task of WorkshopConverter class is to implement Create() method.

olsavchenk commented on OutOfSchool/OutOfSchool.Redis/IReadWriteCacheService.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    Task<string> ReadAsync(string key);
+
+    public Task WriteAsync(
```
(Code style) Please remove “public” as it is public anyway

HlibBondarev replied:
Ok, I've done it

olsavchenk commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/IncompletedWorkshops/WorkshopWithContactsDto.cs:
```diff
@@ -0,0 +1,14 @@
+using OutOfSchool.BusinessLogic.Util.JsonTools;
+
+namespace OutOfSchool.BusinessLogic.Models.Workshops.IncompletedWorkshops;
+public class WorkshopWithContactsDto : WorkshopWithDescriptionDto
```
(Code style) Please add a blank line between the namespace and the class. And in other places too (see WorkshopWithDescriptionDto, WorkshopWithTeachersDto etc.)

HlibBondarev replied:
Ok, I've done it

HlibBondarev committed:
Resolved OleksSavchenko issues:
1) Used an exspression body for CanWrite() and CanConvert() methods.
2) Removed 'public' for Task WriteAsync() method in IReadWriteCacheService interface.
3) Added a blank line between the namespace and the class in WorkshopWithDescriptionDto, WorkshopWithContactsDto and WorkshopWithTeachersDto.

HlibBondarev committed:
Deleted string interpolation in logging message templates.

HlibBondarev committed:
Fixed JsonCreationConverter<T> and CacheService - used ArgumentNullException.ThrowIfNull().
Fixed unit tests in CacheServiceTests.

HlibBondarev committed:
1) Renamed IMememntoService<T> to IDraftStorageService<T> and renamed appropriated service classes and tests.
2) Fixed unit tests in DraftStorageServiceTests.

HlibBondarev committed:
Renamed Memento folder to DraftStorage for tests.

HlibBondarev committed:
1) Renamed MementoController<T> to DraftStorageController<T>.
2) Renamed WorkshopMementoController to WorkshopDraftStorageController.
3) Renamed WorkshopMementoControllerTests to WorkshopDraftStorageControllerTests.

HlibBondarev committed:
Changed RemoveAsync() method of DraftStorageService<T> class.

HlibBondarev committed:
Refactored GetOrAddAsync<T>() method of CacheService class - relocated var returnValue of T type.

HlibBondarev committed:
1) Changed RemoveAsync() method of CacheService class - deleted throwing exception when valueToRemove == null.
2) Changed RestoreAsync() method of CacheService class -  RestoreAsync returned Task<T> instead of Task<T?>.

HlibBondarev committed:
Changed tests in DraftStorageServiceTests class.

HlibBondarev committed:
Deleted redundant directive 'using System' in DraftStorageServiceTests class

HlibBondarev committed:
1) Renamed the Memento folder to DraftStorage in the Service folder of the OutOfSchool.BusinessLogic project.
2) Renamed the appropriate Using directives.

HlibBondarev committed:
Deleted line

HlibBondarev committed:
Added line

HlibBondarev committed:
1) Deleted JSONCreationConverter and WorkshopConverter classes.
2) Removed dependency injection (WorkshopConverter class) from Startup class.
3) Deleted IncompletedWorkshops folder with dtos - WorkshopWithRequiredPropertiesDto, WorkshopWithDescriptionDto, WorkshopWithContactsDto and WorkshopWithTeachersDto.
4) Restored WorkshopBaseDto code from develop branch.
5) Changed WorkshopDraftStorageController code - used IDraftStorageService<WorkshopBaseDto> argument.
6) Changed unit tests in DraftStorageServiceTests and WorkshopDraftStorageControllerTests classes.

HlibBondarev committed:
Deleted line

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Redis/CacheServiceTests.cs:
```diff
@@ -85,14 +86,71 @@ public async Task GetOrAddAsync_WhenDataNotExistsOrExpired_ShouldSaveNewData()
+        {
+            {"ExpectedKey", "ExpectedValue"},
+        };
+        distributedCacheMock.Setup(c => c.Get(It.IsAny<string>()))
```
(Testing wrong data) Please avoid usage of `It.IsAny<>()` whenever possible. Here you know the key thus could specify it in `Setup`. Please do the same in other tests as well

For this specific test it just adds better understanding of the logic. But for some tests it might be crucial.

HlibBondarev replied:
Ok, done. I replaced It.IsAny<> with specific known values.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Redis/CacheServiceTests.cs:
```diff
@@ -85,14 +86,71 @@ public async Task GetOrAddAsync_WhenDataNotExistsOrExpired_ShouldSaveNewData()
+    public async Task ReadAsync_WhenDataExistsInCacheAndNotExpired_ShouldReturnData()
+    {
+        // Arrange
+        var expected = new Dictionary<string, string>()
+        {
+            {"ExpectedKey", "ExpectedValue"},
+        };
```
(Complicated test) Please pay attention what you are testing.

For this specific test (and few next) - sorry, but `expected` variable:
1. It is not "expected" (expected is serialized presentation of this dictionary)
2. It has wrong type - as `ReadAsync` returns just string without extra processing - why not check strings without that serialization / Contains magic?
3. It is hardcoded - why not use Bogus random string?

HlibBondarev replied:
Ok, done.
1) Used Bogus random string for generating data.
2) Removed dictionary.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Redis/CacheServiceTests.cs:
```diff
@@ -85,14 +86,71 @@ public async Task GetOrAddAsync_WhenDataNotExistsOrExpired_ShouldSaveNewData()
+        // Arrange
+        var expected = new Dictionary<string, string>()
+        {
+            {"ExpectedKey", "ExpectedValue"},
```
(Magic strings) `"ExpectedKey"` is used twice. Please think about using const. Same for next tests

HlibBondarev replied:
Ok, done.

Used Bogus random string for generating data.
Removed dictionary.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Redis/CacheServiceTests.cs:
```diff
@@ -85,14 +86,71 @@ public async Task GetOrAddAsync_WhenDataNotExistsOrExpired_ShouldSaveNewData()
+
+        // Assert
+        result.Should().Contain("ExpectedValue");
+        distributedCacheMock.Verify(
```
(Mock calls verification) In general, please prefer `VerifyAll` over specific `Verify(..., Times.Once)`. Former lets you check that absence of excessive `Setup`s

HlibBondarev replied:
Ok, done.
Used VerifyAll() method for tests instead of Verify().

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Services/DraftStorage/DraftStorageServiceTests.cs:
```diff
@@ -0,0 +1,157 @@
+    public async Task RestoreAsync_WhenDraftExistsInCache_ShouldRestoreAppropriatedEntity()
+    {
+        // Arrange
+        var workshopDraft = new WorkshopBaseDto()
```
(Fake data) Test data could and should be randomly generated. See other tests using Faker.

HlibBondarev replied:
Ok, done.
Used Bogus.Facker for creating Workshop draft with random properties.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Services/DraftStorage/DraftStorageServiceTests.cs:
```diff
@@ -0,0 +1,157 @@
+        };
+        var expected = new Dictionary<string, string>()
+            {
+                {"ExpectedKey", "{\"Title\":\"title\",\"Email\":\"myemail@gmail.com\",\"Phone\":\"+380670000000\"}"},
```
Please cosider refactoring (`JsonConvert.SerializeObject` and usage of `const`)

There are two issues:
1. Manual json serialization. Wish a good luck for next developer trying to change `workshopDraft`
2. Magic string - `"ExpectedKey"` - used here and in `Setup` and in `RestoreAsync`. Another "good luck wishes" for changing this test.

HlibBondarev replied:
Ok, done.
Used Bogus random string for generating key and cacheKey instead hardcore string.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.WebApi.Tests/Services/DraftStorage/DraftStorageServiceTests.cs:
```diff
@@ -0,0 +1,157 @@
+        readWriteCacheServiceMock.Verify(
+            c => c.ReadAsync(It.IsAny<string>()),
+            Times.Once);
+        result.Should().BeOfType<WorkshopBaseDto>();
+        Assert.AreEqual(workshopDraft.Title, result.Title);
```
(Code style) `Should()` and `Assert()` in one test. Consider usage of `Should....Which` or `Should...Where` or `BeEquivalentTo`

HlibBondarev replied:
Ok, done. 
Use Should() in all places instead of Assert().

HlibBondarev committed:
Changed tests in CacheServiceTests:
1) Used Bogus random string for generating key and value.
2) Changed It.IsAny<> to specific known values.
3) Used VerifyAll() method for tests.

HlibBondarev committed:
Changed tests in DraftStorageServiceTests and WorkshopDraftStorageControllerTests classes:
1) Used Bogus random string for generating key and cacheKey.
2) Used Bogus.Facker for creating Workshop draft with random properties.
3) Changed It.IsAny<> to specific known values.
4) Used VerifyAll() method for tests.

HlibBondarev committed:
Changed tests in CacheServiceTests - used VerifyAll() method instead of Verify() for tests.

HlibBondarev committed:
Added try-catch block to Create action of WorkshopController.

HlibBondarev committed:
Revers commit - Added try-catch block to Create action of WorkshopController.
